var/spack/repos/builtin/packages/meson/package.py

```diff
@@ -66,6 +66,10 @@ class Meson(PythonPackage):
     depends_on("python@3.7:", when="@0.62.0:", type=("build", "run"))
     depends_on("python@3.6:", when="@0.57.0:", type=("build", "run"))
     depends_on("python@3.5:", type=("build", "run"))
+    # Python 3.12: distutils issues are not resolved yet, see
+    # https://github.com/mesonbuild/meson/issues/7702 and
+    # https://github.com/mesonbuild/meson/compare/master...dnicolodi:meson:rm-distutils
+    depends_on("python@:3.11", type=("build", "run"))
```

Comment:
This is technically untrue: meson itself works fine on python 3.12, however when using:
`meson
py = import('python')
py.find_installation('python3.12')
```
The external python 3.12 executable that meson runs as a subprocess needs to have distutils.

Meson will often fall back to detecting the version of python which it internally has as `sys.executable` but this is not a requirement, and meson-python will *always* specify the exact python program to use, based on the version of python from `pip install` or `python -m build`; this may or may not be the same version of python that was used to install meson, but in environments where there are multiple versions of python it will very often not be the same.

Either way, installing the `py-setuptools` module as is already done on the following two lines is all you need to make this work, and is already done. So I would recommend removing this commit.

Reply:
yup, so better to keep a conflict on meson for now.

Reply:
Can you explain what you mean by this? As far as I can tell, spack already has a fix that makes meson work just fine on python 3.12 and therefore this conflict should not be needed.

Reply:
Sorry, removed my previous comment when I noticed you made a large edit to your initial message.

I added the conflict to be a bit more pragmatic: packages that depend on meson and python usually depend on the same python as meson.

So sure, meson runs fine under python 3.12, but meson.build files that locate the same python will fail.

I could add a comment that technically the conflict should move one level up to packages that use meson and cannot locate python 3.12 as a result, but that's a lot of work, and I assume a fix for meson can be expected soon.

Reply:
> Sorry, removed my previous comment when I noticed you made a large edit to your initial message.

Technically I did not edit my message, I just fixed a buggy single backtick to three backticks which caused most of my message to accidentally end up in a code block. :(

Reply:
Yeah, that pth hack requires the file to be in system site-packages, which won't work with Spack.

Reply:
> I still consider it a bug in meson, since meson now requires setuptools to be installed in a very specific way. 

The bug in meson is that we depend on a non-stdlib module called "distutils", which was dropped from the stdlib. It's available standalone (https://github.com/pypa/distutils) and *also* in setuptools but meson doesn't "require" you to do anything in particular, this is just the workaround that several downstreams are using -- and also the general workaround for thousands of modules with a setup.py file using distutils.

Depending on setuptools is just as much of a bug since it is not in the stdlib either, and one that I don't like for the fact that it in turn breaks Debian.

> That works in general and seems simple enough to go into a patch release.

1. it breaks Debian 
2. I see extremely low gain in adding this hack when there's already a PR that will simply drop the use of distutils altogether. It could be merged already but there was a concern raised about PyPy.

> Yeah, that pth hack requires the file to be in system site-packages, which won't work with Spack.

Oh. :( This happens because processing .pth files has to use os.listdir() on the site directory. (I have no idea how spack lays out its python install.)

Reply:
> I have no idea how spack lays out its python install

Each package in its own dir. Same principle is used by nix / guix / gentoo prefix, I guess

Reply:
Let me try to summarize:

- Meson's detection logic for Python 3.12 only works under conditions that don't apply to Spack. (namely `setuptools` being installed as a "system package")
- There won't be a patch release of Meson that fixes this issue
- Meson itself runs fine under Python 3.12

If so, I'd rather put the upperbound on python in the meson package, instead of going over all dependents of meson that happen to also depend on python, and having to duplicate this conflict there.

Reply:
> Each package in its own dir, so python and python packages are in different dirs. Same principle is used by nix / guix / gentoo prefix, I guess


nix yes, guix yes.

Gentoo Prefix is a normal system that happens to be installed to a custom prefix, it doesn't attempt to overflow the maximum size of a process startup environment via redefinition of libc through shell script wrappers setting LF_LIBRARY_PATH.

(Gentoo Prefix doesn't attempt any saner methods of handling per-package environments, like per-package RPATH. It is "just" a single prefix and everything is installed into it.)

Reply:
> There won't be a patch release of Meson that fixes this issue

There will be a patch release but it will use the existing PR rather than one that adds an `import setuptools`. :)

Reply:
All this being said... I think that you can then remove the runtime dependency on py-setuptools? It seems to have been mistakenly added under the belief that it would solve a problem that it doesn't in fact solve.

Reply:
I've moved the conflict to `MesonPackage` anyways, so the conflict is now duplicated for each and every package that uses meson as a build system.

Reply:
Note that numpy and scipy both use meson (as of #40057) but don't extend MesonPackage. So this conflict may make more sense in the meson package.